fix: revamp of the integration checklist (#2175)
mendersoftware/mender-docs: 03.Client-installation/08.Integration-checklist/docs.md

    category: docs
---

To ensure that the necessary components for Mender are properly integrated, you should use this checklist to verify each of them in turn. You can run this checklist after you have successfully built all components and correctly booted the device.

Please note that the full checklist is required only for doing full rootfs updates.

## Introduction

This checklist will verify some key functionality aspects of the Mender integration. It will verify that:

* The Mender service launches properly
* Bootloader environment tools are present on the device
* Identify A/B partitions
* Confirm OS switch using bootloader variables
* Identify the transitions state 
  * a) Confirm behavior for the 'successful update' case
  * b) Confirm behavior for the 'failed update and rollback' case

! **Please note** During these steps we will be switching between the two partitions without downloading new content to them. 
! If you don't have a bootable OS on the second partition, nothing will boot once we do the switch.

> ! One way to overcome this is to do one regular rootfs update after the initial flashing before proceeding with the verification.

Reviewer: Aren't these instructions a bit circular? We are trying to verify that rootfs updates work, but then we are asking them to use what we are verifying before starting.

Granted, I suppose this issue isn't new. I'm not sure what's the best approach here though... Chicken and egg problem. Any ideas?

Author: Yes!
But that's the best bad way I have to solve it at the moment. :(

Author: The solution is to ensure that generated images which are flashed directly to the device have duplicated content on both partitions.
I've experienced:
* the login credentials sometimes not working when manually switching
* the content of the second partition non existing 
    * this happened a while ago not for this test, for a yocto image. Apparently we deliberately removed the duplication of the second partition content. (how I understood Lluis, also quite a while ago)

Reviewer: > the content of the second partition non existing
> 
>     * this happened a while ago not for this test, for a yocto image. Apparently we deliberately removed the duplication of the second partition content. (how I understood Lluis, also quite a while ago)

Yeah, I remember.. I can't recall the exact reason right now, but I remember that it wasn't for nothing, there was some improvement because of it.

I don't really have any better ideas for this section either, so I guess we can just leave it like this for now.

## The Mender service launches properly

> !!! If you don't need full rootfs updates you only need to verify this step

Reviewer: Slightly conflicting with the above instruction:

> Please note that the full checklist is required only for doing full rootfs updates.

Maybe we can just remove that sentence above altogether and wait until they get down to this step right here to mention it?

Author: Removing: `Please note that the full checklist is required only for doing full rootfs updates.` from above.

mender-client is the userspace application responsible for executing the updates.
It runs as a systemd service.

> To verify it's successfully running run the following commands and confirm the output:

Reviewer 2: This actually just checks if the process is correctly started by `systemd`

Reviewer 2: ```suggestion
To verify the correct installation of the service, run the following commands and confirm the output:
```

```
systemctl is-active mender-client
# active

systemctl is-enabled mender-client
# enabled
```

> In the remaining verification, we'll manually be executing similar steps to what the mender-client usually does in regular operations to communicate with the bootloader.

Reviewer 2: Taxonomy and casual language.

Reviewer 2: ```suggestion
In the remaining verification, we will manually be executing similar steps to what the Mender Client usually does in regular operations to communicate with the bootloader.
```

> To not have the mender-client interfere with the manual verification it's best to disconnect the device from the internet to prevent an update happening during verification time.

Reviewer 2: ```suggestion
To avoid the Mender Client interfering with the manual verification it is recommended disconnect the device from the internet.
```

## Bootloader environment tools are present on the device

> Verify that the two commands to manipulate the bootloader environment are in the path and are executable. These are used by the mender-client to communicate with the bootloader. 

Reviewer 2: ```suggestion
Verify that the two commands to manipulate the bootloader environment are in the path and are executable. These are used by the Mender Client to communicate with the bootloader. 
```

> Depending on the bootloader you're using these might be GRUB or uboot specific.

Reviewer 2: Casual language, clarification.

Reviewer 2: ```suggestion
The executables which the Mender Client expects are bootloader specific. By default, `GRUB` and `uboot` implementations are supported.
```

Please run the commands below:

* For GRUB:
    * `grub-mender-grubenv-print`
    * `grub-mender-grubenv-set`
* For U-Boot:
    * `fw_printenv`
    * `fw_setenv`

> If you get errors when calling the `set` command without arguments, that's ok.

Reviewer 2: ```suggestion
Errors upon calling the `set` command without arguments are expected and can be ignored, this step only checks the executable form.
```

We will test the setting of variables in the upcoming steps.

For the remaining steps, the GRUB CLI tools will be used, but the verification steps are equivalent when using the uboot tools.

<!--AUTOVERSION: "prior to 4.0 %"/ignore-->

> ! In Yocto releases prior to 4.0 kirkstone, the names of the GRUB tools were the same as the U-Boot tools. Make sure to take this into account in the remaining examples on this page.

Author: Nope :(
Don't know what this is about, just fixed the formatting.

## Identify A/B partitions

Redundant (A/B) partitioning is a requirement for full rootfs updates.

> These steps will identify the partitions and check if they align with what is in mender-client configuration.

Reviewer 2: Taxonomy

Reviewer 2: ```suggestion
These steps will identify the partitions and check if they align with what is in the Mender Client configuration.
```

Please note that depending on the types of disks you're using the actual device names can vary.

> Check the partitions set in mender.conf to identify rootfs partitions:

Reviewer 2: ```suggestion
Check the partitions set in `mender.conf` to identify rootfs partitions:
```

``` bash
cat /var/lib/mender/mender.conf | grep RootfsPart
# "RootfsPartA": "/dev/nvme0n1p2"

> # "RootfsPartB": "/dev/nvme0n1p3"

Reviewer: Are the `#` characters intentional? They most likely won't be there, right? At least not for a working configuration.

Author: Fair, it's an attempt to make it copy pastable and not have the output make the commands break.
Would you find this clearer:

```
#  Output:
# "RootfsPartA": "/dev/nvme0n1p2"
# "RootfsPartB": "/dev/nvme0n1p3"
```

Reviewer: It's clearer, yes! 👍

```

> ! By default mender-client look for [configuration at two locations](../07.Configuration-file/docs.md). One of those is `/var/lib/mender/mender.conf` which is - in the default case - a link to `/data/mender/mender.conf` and doesn't get overwritten during the rootfs update. We recommend keeping the backup of the `RootfsPartA/B` settings in `/var/lib/mender/mender.conf` as it's very rare that you need to change partition names as a result of an update.

Reviewer 2: Taxonomy, wording.

Reviewer 2: ```suggestion
! By default the Mender Client looks for [configuration in two locations](../07.Configuration-file/docs.md). One of those is `/var/lib/mender/mender.conf` which is - in the default case - a link to the persistent partition `/data/mender/mender.conf` and doesn't get overwritten during the rootfs update. We recommend keeping the backup of the `RootfsPartA/B` settings in `/var/lib/mender/mender.conf` as it is very rare that you need to change partition names as a result of an update.
```

Identify the currently active partition with the following command:

``` bash
mount | grep 'on /'

> #/dev/nvme0n1p3 on / type ext4 ... ...

Reviewer: `#` here also. If you're trying to separate command from output, then perhaps it's better to split the code blocks.

I won't point out more of these.

Author: Concluding, will address them whole in the upper comment.

```

> If the device listed is an ambiguous device, such as `/dev/root`, you can use an alternative method for verifying it. If you call the following series of commands:

Reviewer 2: ```suggestion
On some devices the rootfs is not listed as a block device but as `/dev/root` or similar. You can use an alternative method for verifying it, by calling the following series of commands:
```

```bash
stat -c %D /
# 10303
stat -c %t%02T /dev/nvme0n1p3
# 10303
```

The output of the two commands should be identical. This verifies that the correct partition is mounted as the root device when partition A is active.

At the end of this step, we should have successfully identified the active partition (`nvme0n1p3` in the example) and the inactive partition (`nvme0n1p2`). We will use this info when proceeding with the test.

##  Confirm OS switch using bootloader variables

We will confirm the bootloader can read the environment and is behaving correctly by manually switching to the inactive partition.

In the previous step, we identified the currently running partition to be `nvme0n1p3` and the inactive one `nvme0n1p2`.

> Let's manually set the bootloader variable so it boots from the inactive partition on the next reboot.

Reviewer 2: ```suggestion
Set the bootloader variables manually so it boots from the currently inactive partition on the next reboot.
```

``` bash

> grub-mender-grubenv-set mender_boot_part 2

Author: Expanded with a section on this

> grub-mender-grubenv-set mender_boot_part_hex 2

Reviewer 2: Additionally, why `mender_boot_part` and `mender_boot_part_hex`? Either needs an explanation, or a link to further documentation.

Author: Added a section on this as well

reboot
```

After the device boots up verify that you are indeed running on the expected partition:

``` bash
mount | grep 'on /'
#/dev/nvme0n1p2 on / type ext4 ... ...
```

> Once that is complete repeat the step to return to the previously active partition:

Reviewer 2: Minor language improvement.

Reviewer 2: ```suggestion
After completion, return to the previously active partition by adapting the previous steps:
```

``` bash
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3
reboot
```

## Transition state

In the [Mender state machine workflow](../../06.Artifact-creation/08.Create-a-custom-Update-Module/docs.md#the-state-machine-workflow) the transitional state for the bootloader starts with `ArtifactReboot` and ends with either `ArtifactCommit` or `ArtifactFailure`.

> For this verification, this is a period during which the bootloader environmental variables are expected to change - either by mender-client or the bootloader itself - and the bootloader is expected to enact conditional logic. For comparison, in the non-transition state, the bootloader variables are stable. 

Reviewer 2: Taxonomy

Reviewer 2: ```suggestion
For this verification, this is a period during which the bootloader environmental variables are expected to change - either by the Mender Client or the bootloader itself - and the bootloader is expected to enact conditional logic. For comparison, in the non-transition state, the bootloader variables are stable. 
```

To notify the bootloader about the switch to the transitional state we will set the following variables:

``` bash
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0
```

Setting `upgrade_available` to `1` has multiple side effects:

* the bootloader will start incrementing `bootcount` for every new boot attempt

> * mender-client and the bootloader know there is a partitioning switch taking place

Reviewer 2: Taxonomy

Reviewer 2: ```suggestion
* the Mender Client and the bootloader know there is a partitioning switch taking place
```

    * they can make decisions on when to rollback or commit
* returning the `upgrade_available` to `0` represents the conclusion of the transition state

### Confirm behavior for the successful update case

> Once you set the variables from above, confirm your current active partition and reboot:

Reviewer 2: Rewording suggestion.

Reviewer 2: ```suggestion
As explained on the variables in the previous paragraph, test a full active partition switch including reboot:
```

``` bash
# In the normal update process, at this point mender-client just
# concluded streaming the new version to the inactive partition

# We are currently running the active partition
# Identify active partition
mount | grep 'on /'
#/dev/nvme0n1p3 on / type ext4 ... ...

# Start the transition state
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0

# Switch to the inactive partition
grub-mender-grubenv-set mender_boot_part 2
grub-mender-grubenv-set mender_boot_part_hex 2

reboot
```

After the reboot the partition changed and the bootcount increased:

``` bash
# Identify the active partition
mount | grep 'on /'
#/dev/nvme0n1p2 on / type ext4 ... ...

grub-mender-grubenv-print bootcount upgrade_available
# bootcount=1
# upgrade_available=1

# All good, concluding the transition state
grub-mender-grubenv-set upgrade_available 0
```

### Confirm behavior for the 'failed update and rollback' case

> The starting steps are similar to the successful case:

Reviewer 2: Rewording suggestion

Reviewer 2: ```suggestion
The process initiation is identical to the "success" form.
```

``` bash
# In the normal update process, at this point mender-client just
# concluded streaming the new version to the inactive partition

# We are currently running the active partition
# Identify the active partition
mount | grep 'on /'
#/dev/nvme0n1p2 on / type ext4 ... ...

# Start the transition state
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0

# Switch to the inactive partition
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3
```

> With the distinction where we're going to mess up the kernel on the inactive partition.

Reviewer 2: Rewording suggestion.

Reviewer 2: ```suggestion
To trigger the rollback mechanism, rename the kernel on the inactive partition to break the boot process.
```

This is equivalent to a use case where the new update contains a faulty kernel.

<!--AUTOVERSION: "vmlinuz-%"/ignore-->
``` bash
mkdir /mnt/inactive-partition
mount /dev/nvme0n1p3 /mnt/inactive-partition
mv /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic.backup

umount /mnt/inactive-partition

reboot
```

The device will attempt to boot, fail and then roll back to booting from the previously working partition.
The bootloader will automatically conclude the transition state in that case:

``` bash
grub-mender-grubenv-print upgrade_available
# upgrade_available=0
```

> ! **Please note** - the trigger for the rollback is the fact that an unplanned second reboot happened during the transition period, not the fact that the bootloader detected a faulty kernel.

Reviewer 2: Rewording suggestion

Reviewer 2: ```suggestion
! **Please note** - the rollback is triggered by an unplanned second reboot happened during the transition period, not the bootloader somehow detecting a faulty kernel.
```

! Any unexpected reboot in the transition period will result in a rollback.

##  Conclusion

If you have verified all the steps you have confirmed your device has the correct partitioning and bootloader integration to do full rootfs updates.

> f you have verified all the steps you have confirmed your device has the correct partitioning and bootloader integration to do full rootfs updates.

Reviewer: Repeated line.

> f you have verified all the steps you have confirmed your device has the correct partitioning and bootloader integration to do full rootfs updates.

Reviewer 2: I guess that line was accidentially commited.

    category: docs
---

To ensure that the necessary components for Mender are properly integrated, you should use this checklist to verify each of them in turn. You can run this checklist after you have successfully built all components and correctly booted the device.

## Introduction

This checklist will verify some key functionality aspects of the Mender integration. It will verify that:

* The Mender service launches properly
* Bootloader environment tools are present on the device
* Identify A/B partitions
* Confirm OS switch using bootloader variables
* Identify the transitions state 
  * a) Confirm behavior for the 'successful update' case
  * b) Confirm behavior for the 'failed update and rollback' case

! **Please note** During these steps we will be switching between the two partitions without downloading new content to them. 
! If you don't have a bootable OS on the second partition, nothing will boot once we do the switch.

> ! One way to overcome this is to do one regular rootfs update after the initial flashing before proceeding with the verification.

Author: Will update the wording.
Don't get the `check the specific board integration` part, can't this be a `dd` snippet?

> ! One way to overcome this is to do one regular rootfs update after the initial flashing before proceeding with the verification.

Author: No follow up. Mrking as resolved.

## The Mender service launches properly

!!! If you don't need full rootfs updates you only need to verify this step

The Mender Client consists of a number of components and configuration files, with the `mender-client` userspace application being the core responsible for executing the updates.
By default, it runs as a systemd service.
To verify the correct installation of the service, run the following commands and confirm the output:

```
systemctl is-active mender-client
# Output:
# active

systemctl is-enabled mender-client
# Output:
# enabled
```

In the remaining verification, we will manually be executing similar steps to what the Mender Client usually does in regular operations to communicate with the bootloader.

> To avoid the Mender Client interfering with the manual verification it is recommended disconnect the device from the internet.

Reviewer: "recommended **to** disconnect"

The executables which the Mender Client expects are bootloader specific. By default, `GRUB` and `uboot` implementations are supported.
Please run the commands below:

* For GRUB:
    * `grub-mender-grubenv-print`
    * `grub-mender-grubenv-set`
* For U-Boot:
    * `fw_printenv`
    * `fw_setenv`

Errors upon calling the `set` command without arguments are expected and can be ignored, this step only checks the executable form.
We will test the setting of variables in the upcoming steps.

For the remaining steps, the GRUB CLI tools will be used, but the verification steps are equivalent when using the uboot tools.

<!--AUTOVERSION: "prior to 4.0 %"/ignore-->
! In Yocto releases prior to 4.0 kirkstone, the names of the GRUB tools were the same as the U-Boot tools. Make sure to take this into account in the remaining examples on this page.

## Identify A/B partitions

Redundant (A/B) partitioning is a requirement for full rootfs updates.
These steps will identify the partitions and check if they align with what is in the Mender Client configuration.

> Please note that depending on the types of storage the actual device names can vary.

Author: Don't want to link to external docs, looks like an unnecessary dependency.
Risks it becoming stale or link broken.

The device name is whatever the device name is, and if someone want's to know more, they can research on their own.

# Output:
# "RootfsPartA": "/dev/nvme0n1p2"
# "RootfsPartB": "/dev/nvme0n1p3"
```

! By default the Mender Client looks for [configuration in two locations](../07.Configuration-file/docs.md). One of those is `/var/lib/mender/mender.conf` which is - in the default case - a link to the persistent partition `/data/mender/mender.conf` and doesn't get overwritten during the rootfs update. We recommend keeping the backup of the `RootfsPartA/B` settings in `/var/lib/mender/mender.conf` as it is very rare that you need to change partition names as a result of an update.

Identify the currently active partition with the following command:

``` bash
mount | grep 'on /'
# Output:
#/dev/nvme0n1p3 on / type ext4 ... ...
```

On some devices the rootfs is not listed as a block device but as `/dev/root` or similar. You can use an alternative method for verifying it, by calling the following series of commands:

```bash
stat -c %D /
# Output:
# 10303
stat -c %t%02T /dev/nvme0n1p3
# Output:
# 10303
```

The output of the two commands should be identical. This verifies that the correct partition is mounted as the root device when partition A is active.

> At the end of this step, we should have successfully identified the active partition (`nvme0n1p3` in the example) and the inactive partition (`nvme0n1p2`). We will use this info when proceeding with the test.

Reviewer 3: ```suggestion
At the end of this step, we should have successfully identified the active partition (`nvme0n1p3` in the example) and the inactive partition (`nvme0n1p2`). This information will be used later when we perform the switch test between partitions.
```

Author: Expanded with UUID example.

##  Confirm OS switch using bootloader variables

!!! We identified edge cases in certain u-boot board integrations which lead to the introduction of the `mender_boot_part_hex` variable.

> !!! To make the verification steps generally applicable, we change both variables in the steps event though they aren't both used in all cases.

Reviewer: "...**even** though..."

``` bash
mount | grep 'on /'
# Output:
#/dev/nvme0n1p2 on / type ext4 ... ...
```

After completion, return to the previously active partition by adapting the previous steps:

``` bash
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3
reboot
```

## Transition state

In the [Mender state machine workflow](../../06.Artifact-creation/08.Create-a-custom-Update-Module/docs.md#the-state-machine-workflow) the transitional state for the bootloader starts with `ArtifactReboot` and ends with either `ArtifactCommit` or `ArtifactFailure`.

For this verification, this is a period during which the bootloader environmental variables are expected to change - either by the Mender Client or the bootloader itself - and the bootloader is expected to enact conditional logic. For comparison, in the non-transition state, the bootloader variables are stable. 

To notify the bootloader about the switch to the transitional state we will set the following variables:

``` bash
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0
```

Setting `upgrade_available` to `1` has multiple side effects:

> * the bootloader will start incrementing `bootcount` for every new boot attempt

Reviewer 3: ```suggestion
* The bootloader will start incrementing by 1 the `bootcount` variable for every new boot attempt
```

> * the Mender Client and the bootloader know there is a partitioning switch taking place

Reviewer 3: ```suggestion
* The Mender Client and the bootloader become aware that there is a partitioning switch taking place
```

>     * they can make decisions on when to rollback or commit

Reviewer 3: ```suggestion
    * They can make decisions on when to rollback or commit
```

> * returning the `upgrade_available` to `0` represents the conclusion of the transition state

Reviewer 3: ```suggestion
* Returning the `upgrade_available` value to `0` implies the conclusion of the transition state
```

# We are currently running the active partition
# Identify the active partition
mount | grep 'on /'
# Output:
#/dev/nvme0n1p2 on / type ext4 ... ...

# Start the transition state
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0

# Switch to the inactive partition
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3
```

To trigger the rollback mechanism, rename the kernel on the inactive partition to break the boot process.
This is equivalent to a use case where the new update contains a faulty kernel.

<!--AUTOVERSION: "vmlinuz-%"/ignore-->
``` bash
mkdir /mnt/inactive-partition
mount /dev/nvme0n1p3 /mnt/inactive-partition
mv /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic.backup

umount /mnt/inactive-partition

reboot
```

The device will attempt to boot, fail and then roll back to booting from the previously working partition.
The bootloader will automatically conclude the transition state in that case:

``` bash
grub-mender-grubenv-print upgrade_available
# Output:
# upgrade_available=0
```

> ! **Please note** - the rollback is triggered by an unplanned second reboot happened during the transition period, not the bootloader somehow detecting a faulty kernel.

Reviewer 3: ```suggestion
! **Please note** - the rollback is triggered by an unplanned second reboot that takes place during the transition period, and it is not a result of the bootloader detecting a faulty kernel in any way.
```

! Any unexpected reboot in the transition period will result in a rollback.

##  Conclusion

> If you have verified all the steps you have confirmed your device has the correct partitioning and bootloader integration to do full rootfs updates.

Reviewer 3: ```suggestion
If you have verified all the steps, then you have confirmed your device has the correct partitioning and bootloader integration to do full rootfs updates using Mender.
```

# enabled
```

In the remaining verification, we will manually be executing similar steps to what the Mender Client usually does in regular operations to communicate with the bootloader.

To avoid the Mender Client interfering with the manual verification it is recommended to disconnect the device from the internet.

## Bootloader environment tools are present on the device

Verify which of the two commands to manipulate the bootloader environment are executable and available in the path. 
These are used by the `mender-client` to set the bootloader environment.

The executables which the Mender Client expects are bootloader specific. By default, `GRUB` and `uboot` implementations are supported.
Please run the commands below:

* For GRUB:
    * `grub-mender-grubenv-print`
    * `grub-mender-grubenv-set`
* For U-Boot:
    * `fw_printenv`
    * `fw_setenv`

Errors upon calling the `set` command without arguments are expected and can be ignored, this step only checks the executable form.
We will test the setting of variables in the upcoming steps.

For the remaining steps, the GRUB CLI tools will be used, but the verification steps are equivalent when using the uboot tools.

> !! PARTUUID feature hasn't been tested for u-boot

Reviewer 2: Maybe "the PARTUUID feature ..."

<!--AUTOVERSION: "prior to 4.0 %"/ignore-->
! In Yocto releases prior to 4.0 kirkstone, the names of the GRUB tools were the same as the U-Boot tools. Make sure to take this into account in the remaining examples on this page.

## Identify A/B partitions

Redundant (A/B) partitioning is a requirement for full rootfs updates.
These steps will identify the partitions and check if they align with what is in the Mender Client configuration (`/var/lib/mender/mender.conf`).

! By default the Mender Client looks for [configuration in two locations](../07.Configuration-file/docs.md). One of those is `/var/lib/mender/mender.conf` which is - in the default case - a link to the persistent partition `/data/mender/mender.conf` and doesn't get overwritten during the rootfs update. We recommend keeping the backup of the `RootfsPartA/B` settings in `/var/lib/mender/mender.conf` as it is very rare that you need to change partition names as a result of an update.

> Please note that the content can vary depending  storage actual device names or if you're using PARTUUIDs.

Reviewer 2: double blank between "depending" and "storage"

We will confirm the bootloader can read the environment and is behaving correctly by manually switching to the inactive partition.

In the previous step, we identified the currently running partition to be `nvme0n1p2` and the inactive one `nvme0n1p3`.
Set the bootloader variables manually so it boots from the currently inactive partition on the next reboot.

``` bash
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3
reboot
```

After the device boots up verify that you are indeed running on the expected partition:

``` bash
mount | grep 'on /'
# Output:
#/dev/nvme0n1p3 on / type ext4 ... ...
```

After completion, return to the previously active partition by adapting the previous steps:

``` bash
grub-mender-grubenv-set mender_boot_part 2
grub-mender-grubenv-set mender_boot_part_hex 2
reboot
```

## Transition state

In the [Mender state machine workflow](../../06.Artifact-creation/08.Create-a-custom-Update-Module/docs.md#the-state-machine-workflow) the transitional state for the bootloader starts with `ArtifactReboot` and ends with either `ArtifactCommit` or `ArtifactFailure`.

> For this verification, this is a period during which the bootloader environmental variables are expected to change - either by the Mender Client or the bootloader itself - and the bootloader is expected to enact conditional logic. For comparison, in the non-transition state, the bootloader variables are stable. 

Reviewer 3: ```suggestion
During this verification process, it is expected that the bootloader's environment variables will experience changes - either by the Mender Client or the bootloader itself - and the bootloader is expected to enact conditional logic. For comparison, in the non-transition state, the bootloader's variables remain unchanged. 
```

> To notify the bootloader about the switch to the transitional state we will set the following variables:

Reviewer 3: ```suggestion
To notify the bootloader about the switch to the transitional state, we will set the following variables:
```

    * They can make decisions on when to rollback or commit
* Returning the `upgrade_available` to `0` marks the end of transition state

### Confirm behavior for the successful update case

As explained on the variables in the previous paragraph, test a full active partition switch including reboot:

``` bash
# In the normal update process, at this point the Mender Client just
# concluded streaming the new version to the inactive partition

# We are currently running the active partition
# Identify active partition
mount | grep 'on / '
# Output:
#/dev/nvme0n1p2 on / type ext4 ... ...

# Start the transition state
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0

# Switch to the inactive partition
grub-mender-grubenv-set mender_boot_part 3
grub-mender-grubenv-set mender_boot_part_hex 3

reboot
```

After the reboot the partition changed and the bootcount increased:

``` bash
# Identify the active partition
mount | grep 'on / '
# Output:
#/dev/nvme0n1p3 on / type ext4 ... ...

grub-mender-grubenv-print bootcount upgrade_available
# Output:
# bootcount=1
# upgrade_available=1
```

> This is as expected, we can conclude the transition state and confirm the variables remain stable:

Reviewer 3: ```suggestion
This is as expected, we can conclude the transition state and confirm the variables remain unchanged:
```

# upgrade_available=0
# mender_boot_part_hex=3
```

### Confirm behavior for the 'failed update and rollback' case

The process initiation is identical to the "success" form.

``` bash
# In the normal update process, at this point the Mender Client just
# concluded streaming the new version to the inactive partition

# We are currently running the active partition
# Identify the active partition
mount | grep 'on / '
# Output:
#/dev/nvme0n1p3 on / type ext4 ... ...

# Start the transition state
grub-mender-grubenv-set upgrade_available 1
grub-mender-grubenv-set bootcount 0

# Switch to the inactive partition
grub-mender-grubenv-set mender_boot_part 2
grub-mender-grubenv-set mender_boot_part_hex 2
```

To trigger the rollback mechanism, rename the kernel on the inactive partition to break the boot process.
This is equivalent to a use case where the new update contains a faulty kernel.

<!--AUTOVERSION: "vmlinuz-%"/ignore-->
``` bash
mkdir /mnt/inactive-partition
mount /dev/nvme0n1p2 /mnt/inactive-partition
mv /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic.backup
umount /mnt/inactive-partition

reboot
```

> The device will attempt to boot, fail and then roll back to booting from the previously working partition.

Reviewer 3: ```suggestion
The device will attempt to boot; however, it will fail and trigger a rollback to booting from the previously working partition.
```

The bootloader will automatically conclude the transition state in that case:

``` bash
grub-mender-grubenv-print upgrade_available
# Output:
# upgrade_available=0
```

And the active partition is still the one we started with.

``` bash
# Identify the active partition
mount | grep 'on / '
# Output:
#/dev/nvme0n1p3 on / type ext4 ... ...
```

> You can return your kernel back to normal again:

Reviewer 3: ```suggestion
Now, you can proceed and return your kernel back to normal again:
```

Author: Verbose, rejecting

<!--AUTOVERSION: "vmlinuz-%"/ignore-->
``` bash
mkdir /mnt/inactive-partition
mount /dev/nvme0n1p2 /mnt/inactive-partition
mv /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic /mnt/inactive-partition/boot/vmlinuz-5.13.0-35-generic.backup
umount /mnt/inactive-partition
```

! **Please note** - the rollback is triggered by an unplanned second reboot that takes place during the transition period, and it is not a result of the bootloader detecting a faulty kernel in any way.

##  Conclusion

> If you have verified all the steps you have confirmed that your device has the correct partitioning and bootloader integration to do full rootfs updates using Mender.

Reviewer 3: ```suggestion
If you have successfully followed and verified all the steps, you can confirm that your device has the appropriate partitioning and bootloader integration for complete rootfs updates using Mender.
```
